quickwit/quickwit-search/src/leaf.rs

```diff
@@ -879,6 +877,130 @@ impl CanSplitDoBetter {
         }
     }
 
+    /// This function tries to detect upfront which splits contain the top n hits and convert other
+    /// split searches to count only searches. It also optimizes split order.
+    ///
+    /// Returns the search_requests with their split.
+    fn optimize(
+        &self,
+        request: Arc<SearchRequest>,
+        mut splits: Vec<SplitIdAndFooterOffsets>,
+    ) -> Result<Vec<(SplitIdAndFooterOffsets, SearchRequest)>, SearchError> {
+        self.optimize_split_order(&mut splits);
+        // TODO: we maybe want here some deduplication + Cow logic
+        let mut split_with_req = splits
+            .into_iter()
+            .map(|split| (split, (*request).clone()))
+            .collect::<Vec<_>>();
+
+        if request.aggregation_request.is_some() || request.search_after.is_some() {
+            return Ok(split_with_req);
+        }
+        let query_ast: QueryAst = serde_json::from_str(request.query_ast.as_str())
+            .map_err(|err| SearchError::InvalidQuery(err.to_string()))?;
+
+        // TODO: Update the logic to handle start_timestamp end_timestamp ranges
+        if request.start_timestamp.is_some() || request.end_timestamp.is_some() {
+            return Ok(split_with_req);
+        }
+
+        if !matches!(query_ast, QueryAst::MatchAll) {
+            return Ok(split_with_req);
+        }
```

Comment:
this code is needlessly complicated.

I suggest we add a function outside of this 

```
/// Returns true if the query is a all query, with no range query,
/// nor search_after, nor aggregation.
fn is_simple_all_query(search_request: &SearchRequest) -> bool {
    if search_request.aggregation_request.is_some() {
        return false;
    }

   if search_request.search_after.is_some() {
       return false;
   }

    // TODO: Update the logic to handle start_timestamp end_timestamp ranges
    if search_request.start_timestamp.is_some() || search_request.end_timestamp.is_some() {
        return false;
    }

    let Ok(query_ast) = serde_json::from_str(&search_request.query_ast) else {
        return false;
    };

    matches!(query_ast, QueryAst::MatchAll)
}
```

We can then replace this block by

```
        if !is_simple_all_query(&*request) {
            // no optimization opportunity here.
            return Ok(splits
                .into_iter()
                .map(|split| (split, (*request).clone()))
                .collect::<Vec<_>>());
        }
```

The `count_required_splits` can then operate directly on `splits` rather than `split_with_req`.